openedx/openedx-platform
# SE-2010 Allows XBlocks to order columns for the Problem Response Report

Allows XBlocks to determine the sorting order for columns in the `generate_report_data` by returning an `OrderedDict`.  This PR alters the report logic to respect that ordering, if provided.  Defaults to the previous behaviour of sorting columns alphabetically, if no order was specified.

WIP: report format is being reviewed for suitability by the client.

**JIRA tickets**: [OSPR-4098](https://openedx.atlassian.net/browse/OSPR-4098)

**Dependencies**: https://github.com/edx/edx-ora2/pull/1350

**Screenshots**:

No change to UI.

**Merge deadline**: None

**Sandbox**

https://pr23109.sandbox.opencraft.hosting
https://studio.pr23109.sandbox.opencraft.hosting

**Example data**

[example CSV](https://github.com/edx/edx-platform/files/4385660/OpenCraft_ORA2_2020_1_student_state_from_block-v1_OpenCraft%2BORA2%2B2020_1%2Btype%40chapter%2Bblock%4096abd57ebe08457cbc545a21f66b3240_2020-03-26-0737.csv.txt)

This file was generated from the sandbox instance [Open Response Assessment Test course](https://pr23109.sandbox.opencraft.hosting/courses/course-v1:OpenCraft+ORA2+2020_1/course/) using the following setup:

* `honor` user submitted a text response.  It was graded by `audit` and `verified`, and `honor` graded those users' responses too, and so her submission's final scores are available.  She also left feedback on the assessments.
* `audit` user submitted a file response.  It was graded by `honor`, and so the report shows one assessment row for this submission.  The submission is awaiting one more assessment before scores can be given.
* `verified` user submitted a text+file response.  It was graded by `honor` and `audit`, and so the report shows 2 assessment rows for this submission.  The submission is waiting for `verified` to do one more assessment before scores can be given.
* `professional` user has submitted a text+file response.  No grading has been given or received, and so assessment information is available in the report.
* `another` user has saved a text+file response, but has not submitted. No grading has been given or received, and so assessment information is available in the report.
* `staff` user has not submitted or saved any responses, and so is not shown in the report.

**Testing instructions**:

1. In Studio, create an Open Response Assessment unit with peer assessment steps.
1. Enroll at least 2 users in the course. Have them complete the ORA problem.
1. As each user moves through the peer assessment process, run the next step and watch how the generated data changes.
1. As a course staff user, visit the Instructor Dashboard and scroll down to the "Select a section or problem" button.
1. Select the section (or subsection, vertical, or block) that contains the ORA2 problem, and click "Download a CSV of problem responses".
1. View the resulting report, and note that the report renders its columns in the order specified by the ORA2 XBlock.

**Author notes and concerns**:

1. In my devstack, I had to add this to `lms/envs/devstack.py` in order to be able to download any reports generated from the Instructor Dashboard:

   ```python
   GRADES_DOWNLOAD['STORAGE_KWARGS'] = None
   ```
1. For the record: The client originally wanted to see the cohort in this report, if cohorts are enabled for the course.
   The project is out of budget, but we proposed adding `cohort` to the [`XBlockUserState`](https://github.com/edx/edx-user-state-client/blob/56aa5f508d956b9338ed6a5dc522c47a70ee9543/edx_user_state_client/interface.py#L25). This would have add the cohort name to Problem Response reports for all XBlock types.

**Reviewers**
